test/Microsoft.ML.Tests/MemoryCollectionTests.cs

```diff
@@ -0,0 +1,209 @@
+                ILearningPipelineDataStep output = (ILearningPipelineDataStep)collection.ApplyStep(null, experiment);
+
+                Assert.NotNull(output.Data);
+                Assert.NotNull(output.Data.VarName);
+                Assert.Null(output.Model);
+            }
+        }
+
+        [Fact]
+        public void CanSuccessfullyEnumerated()
+        {
+            var collection = MemoryCollection.Create(new List<Input>() {
+                new Input { Number1 = 1, String1 = "1" },
+                new Input { Number1 = 2, String1 = "2" },
+                new Input { Number1 = 3, String1 = "3" }
+            });
+
+            using (var environment = new TlcEnvironment())
+            {
+                Experiment experiment = environment.CreateExperiment();
+                ILearningPipelineDataStep output = collection.ApplyStep(null, experiment) as ILearningPipelineDataStep;
+
+                experiment.Compile();
+                collection.SetInput(environment, experiment);
+                experiment.Run();
+
+                IDataView data = experiment.GetOutput(output.Data);
+                Assert.NotNull(data);
+
+                using (var cursor = data.GetRowCursor((a => true)))
+                {
+                    var IDGetter = cursor.GetGetter<float>(0);
+                    var TextGetter = cursor.GetGetter<DvText>(1);
+
+                    Assert.True(cursor.MoveNext());
+
+                    float ID = 0;
+                    IDGetter(ref ID);
+                    Assert.Equal(1, ID);
+
+                    DvText Text = new DvText();
+                    TextGetter(ref Text);
+                    Assert.Equal("1", Text.ToString());
+
+                    Assert.True(cursor.MoveNext());
+
+                    ID = 0;
+                    IDGetter(ref ID);
+                    Assert.Equal(2, ID);
+
+                    Text = new DvText();
+                    TextGetter(ref Text);
+                    Assert.Equal("2", Text.ToString());
+
+                    Assert.True(cursor.MoveNext());
+
+                    ID = 0;
+                    IDGetter(ref ID);
+                    Assert.Equal(3, ID);
+
+                    Text = new DvText();
+                    TextGetter(ref Text);
+                    Assert.Equal("3", Text.ToString());
+
+                    Assert.False(cursor.MoveNext());
+                }
+            }
+        }
+
+        [Fact]
+        public void CanTrain()
+        {
+            var pipeline = new LearningPipeline();
+            var data = new List<IrisData>() {
+                new IrisData { SepalLength = 1f, SepalWidth = 1f ,PetalLength=0.3f, PetalWidth=5.1f, Label=1},
+                new IrisData { SepalLength = 1f, SepalWidth = 1f ,PetalLength=0.3f, PetalWidth=5.1f, Label=1},
+                new IrisData { SepalLength = 1.2f, SepalWidth = 0.5f ,PetalLength=0.3f, PetalWidth=5.1f, Label=0}
+            };
+            var collection = MemoryCollection.Create(data);
```

Comment:
to make it easier for user.

Tom: "Just an observation... if we had instead structured this as a static utility method somewhere, then we could avoid having the double-specification of the Input class, as it would have been inferred by the compiler.

So: if we had a input type MyAwesomeInputType, then instead of MemoryCollection<MyAwesomeInputType>(new MyAwesomeInputType[] {...}), we would have MemoryCollection.Create(new MyAwesomeInputType[] {...}) since the compiler could have done the work of inferring the type and whatnot."

---
In reply to: [187397737](https://github.com/dotnet/machinelearning/pull/106#discussion_r187397737) [](ancestors = 187397737)

Reply:
Consistency might also make things easier for the user. It might be helpful if all the approaches to bringing data into the pipeline follow similar conventions. Another question is whether we want to include "Loader" here as well to emphasize this is loading data.

Reply:
I don't insist on this BTW. Just thought it was nice. It also has the side benefit of allowing us avoid exposing the implementation class as part of the API.

---
In reply to: [187400552](https://github.com/dotnet/machinelearning/pull/106#discussion_r187400552) [](ancestors = 187400552,187397737)